### ISSUE TITLE
+= is not annotated

### ISSUE THREAD

**Issue by jwallwork23** [firedrakeproject/firedrake#1860]
In Firedrake, `target += source` is equivalent to `target.assign(target + source)`, provided `target` and `source` are from the same space. However, this is not annotated by pyadjoint.

Consider the MFE
```
from firedrake import *
from firedrake_adjoint import *

# Create source and target functions in P1 space
mesh = UnitSquareMesh(1, 1)
P1 = FunctionSpace(mesh, "CG", 1)
source = Function(P1)
source.assign(1.0)                # This gets annotated
target = Function(P1)
# target.assign(target + source)  # This gets annotated
target += source                  # This does not

# Examine tape
tape = get_working_tape()
for i, block in enumerate(tape.get_blocks()):
    print(i, type(block))

# Try to compute gradient of some functional w.r.t. source
J = assemble(target*dx)
compute_gradient(J, Control(source))
```
This gives the following outputs under each approach....

**target.assign(target + source)**
```
0 <class 'firedrake.adjoint.blocks.FunctionAssignBlock'>
1 <class 'firedrake.adjoint.blocks.FunctionAssignBlock'>
```

**target += source**
```
0 <class 'firedrake.adjoint.blocks.FunctionAssignBlock'>
WARNING:root:Adjoint value is None, is the functional independent of the control variable?
Traceback (most recent call last):
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/utils.py", line 245, in verify_reshape
    a.shape = shape
ValueError: cannot reshape array of size 1 into shape (4,)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pm.py", line 20, in <module>
    compute_gradient(J, Control(source))
  File "/home/joe/software/firedrake-adapt/src/pyadjoint/pyadjoint/drivers.py", line 31, in compute_gradient
    grads = [i.get_derivative(options=options) for i in m]
  File "/home/joe/software/firedrake-adapt/src/pyadjoint/pyadjoint/drivers.py", line 31, in <listcomp>
    grads = [i.get_derivative(options=options) for i in m]
  File "/home/joe/software/firedrake-adapt/src/pyadjoint/pyadjoint/control.py", line 51, in get_derivative
    return self.control._ad_convert_type(0., options=options)
  File "/home/joe/software/firedrake-adapt/src/pyadjoint/pyadjoint/tape.py", line 46, in wrapper
    return function(*args, **kwargs)
  File "/home/joe/software/firedrake-adapt/src/firedrake/firedrake/adjoint/function.py", line 128, in _ad_convert_type
    return Function(self.function_space(), val=value)
  File "/home/joe/software/firedrake-adapt/src/firedrake/firedrake/adjoint/function.py", line 22, in wrapper
    init(self, *args, **kwargs)
  File "/home/joe/software/firedrake-adapt/src/firedrake/firedrake/function.py", line 254, in __init__
    val=val, name=name, dtype=dtype)
  File "/home/joe/software/firedrake-adapt/src/firedrake/firedrake/function.py", line 77, in __init__
    self.dat = function_space.make_dat(val, dtype, self.name())
  File "/home/joe/software/firedrake-adapt/src/firedrake/firedrake/functionspaceimpl.py", line 469, in make_dat
    return op2.Dat(self.dof_dset, val, valuetype, name)
  File "<decorator-gen-12>", line 2, in __init__
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/utils.py", line 133, in wrapper
    return f(*args, **kwargs)
  File "<decorator-gen-11>", line 2, in __init__
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/utils.py", line 133, in wrapper
    return f(*args, **kwargs)
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/base.py", line 1358, in __init__
    _EmptyDataMixin.__init__(self, data, dtype, self._shape)
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/base.py", line 1282, in __init__
    self._numpy_data = verify_reshape(data, dtype, shape, allow_none=True)
  File "/home/joe/software/firedrake-adapt/src/PyOP2/pyop2/utils.py", line 249, in verify_reshape
    (np.prod(shape), np.asarray(data).size))
pyop2.exceptions.DataValueError: Invalid data: expected 4 values, got 1!
```

**Comment by wence-** [comment]
I think this should be fixed by moving the annotation of these pointwise expressions directly to `assemble_expressions.py` (in particular `evaluate_expression`)

**Comment by wence-** [comment]
Or you could individually write annotation decorators for all the things in function.py

**Comment by jwallwork23** [comment]
Fixed in #1995
